win32ss/user/ntuser/window.c

```diff
@@ -3352,98 +3352,79 @@ NtUserFindWindowEx(HWND hwndParent,
 }
 
 
-/*
- * @implemented
- */
+/* @implemented */
 PWND FASTCALL UserGetAncestor(PWND Wnd, UINT Type)
 {
-   PWND WndAncestor, Parent;
+    PWND Ret, Parent, pwndNode, pwndMessage;
+    PDESKTOP pDesktop = Wnd->head.rpdesk;
+    PWND pwndDesktop = pDesktop->pDeskInfo->spwnd;
 
-   if (UserHMGetHandle(Wnd) == IntGetDesktopWindow())
-   {
-      return NULL;
-   }
+    if (Wnd == pwndDesktop)
+        return NULL;
 
-   switch (Type)
-   {
-      case GA_PARENT:
-         {
-            WndAncestor = Wnd->spwndParent;
-            break;
-         }
+    pwndMessage = pDesktop->spwndMessage;
+    if (Wnd == pwndMessage)
+        return NULL;
 
-      case GA_ROOT:
-         {
-            WndAncestor = Wnd;
-            Parent = NULL;
+    Parent = Wnd->spwndParent;
+    if (!Parent)
+        return NULL;
 
-            for(;;)
+    switch (Type)
+    {
+        case GA_PARENT:
+            return Parent;
+        case GA_ROOT:
+            Ret = Wnd;
+            if (Parent == pwndDesktop)
+                break;
+            do
             {
-               if(!(Parent = WndAncestor->spwndParent))
-               {
-                  break;
-               }
-               if(IntIsDesktopWindow(Parent))
-               {
-                  break;
-               }
-
-               WndAncestor = Parent;
-            }
+                if (Parent == pwndMessage)
+                    break;
+                Ret = Parent;
+                pDesktop = Parent->head.rpdesk;
+                Parent = Parent->spwndParent;
+            } while (Parent != pDesktop->pDeskInfo->spwnd);
             break;
-         }
-
-      case GA_ROOTOWNER:
-         {
-            WndAncestor = Wnd;
-
-            for (;;)
-            {
-               Parent = IntGetParent(WndAncestor);
-
-               if (!Parent)
-               {
-                  break;
-               }
-
-               WndAncestor = Parent;
-            }
+        case GA_ROOTOWNER:
+            for (pwndNode = Parent; pwndNode; pwndNode = IntGetParent(pwndNode))
+                Ret = pwndNode;
             break;
-         }
-
-      default:
-         {
+        default:
             return NULL;
-         }
-   }
+    }
 
-   return WndAncestor;
+    return Ret;
 }
 
-/*
- * @implemented
- */
+/* @implemented */
 HWND APIENTRY
 NtUserGetAncestor(HWND hWnd, UINT Type)
 {
-   PWND Window, Ancestor;
-   HWND Ret = NULL;
+    PWND Window, Ancestor;
+    HWND Ret = NULL;
 
-   TRACE("Enter NtUserGetAncestor\n");
-   UserEnterExclusive();
+    TRACE("Enter NtUserGetAncestor\n");
+    UserEnterExclusive();
```

Comment:
```suggestion
    UserEnterShared();
```
The function only browses the hierarchy of windows, it doesn't (and shouldn't) modify it.

Reply:
Fixed.